### ISSUE TITLE
Cache headers for documents do not factor in the cookie for the store code

### ISSUE THREAD

**Issue by andrewhowdencom** [magento/magento2#11939]
### Preconditions

1. Magento 2.2
2. Varnish 5
3. Nginx 10.3 (Ubuntu)

### Steps to reproduce

1. Visit a 2.2 store, configured as per the specified varnish configuration
2. Observe cache headers for .html documents
3. Switch stores using the `__store` cookie method
4. Observe weird shit returned

More generally, I think that content pages should not be cached. *This is particularly nasty as it has no server side records -- the bug is client side*. 

### Expected result
1. Right page returned 

### Actual result
1. Wrong page returned

The bug appears to have been introduced here:

https://github.com/magento/magento2/commit/61fee9b299f4f7284ff9c8f40ce16533e14d0fdc#diff-69713fbe6922f286a3a2054311655300

I think the authors intent was to ensure pages are cached in Varnish. Indeed, the Varnish documentation notes that section is for handling whether the response should be stored in the cache:

https://book.varnish-software.com/4.0/chapters/VCL_Basics.html#vcl-backend-response

However, the mechanism for setting it uncacheable is to set the `beresp.uncacheable` property to true - not to modify the HTTP cache headers. Indeed, the example seems to indicate that special handling of these cache headers is necessary.

cc @andimov  as the author -- am I wrong heree?

Also, special thanks to @mprototype -- helped me track this one to this point. Fleshing this issue out now, before I start hacking on varnish VCLs.

**Comment by piotrekkaminski** [comment]
Internal ticket MAGETWO-83152

**Comment by andrewhowdencom** [comment]
```
From fa26f4e3c405b09cb4f86874a5ddbed82737b941 Mon Sep 17 00:00:00 2001
From: Andrew Howden <hello@andrewhowden.com>
Date: Wed, 1 Nov 2017 19:32:16 +0100
Subject: [PATCH] AD-HOC fix (Varnish): Restore the behaviour of not caching
 .html pages

With the upgrade to 2.2, the Varnish configuration templates were
updated. They were accordingly repopulated and committed to version
control. However, they included a bug in which content pages (matching
regex /\/$/ and /\.html$/) were shipped to the client browser with the
cache-headers intact.

Varnish caches pages in it's store based on the cache-control
headers[1]. It thus makes sense for cache headers to indicate how long
that page should be cached for -- these are used by Varnish to determine
cache content. It also makes sense that the normal headers are used
rather than specific Varnish headers, as the HTTP protocol was designed
in such a way to be "proxy friendly" -- only relied upon until the next
hop.

In the 2.1.9 release Magento handled this by dropping the cache headers
on all content that did not match a whitelist of "static" content, or
content that had other mechanisms for "cache busting" -- rotating the
content after it's no longer valid by changing the src. However, in the
2.2 branch this was dropped:

5ec0996f37b3681cd1ee7d6d60fed072f3d67336..61fee9b299f4f7284ff9c8f40ce16533e14d0fdc

https://github.com/magento/magento2/commit/5ec0996f37b3681cd1ee7d6d60fed072f3d67336

The goal of this work appears to be to ensure that content pages are
cached in Varnish, but static files are not. The author is reasonably
sure this did not have the intended effect, but rather simply stripped
the "no-cache" overrides from content pages.

The change was made to the "vcl_backend_response" section of the
configuration, which does appear to be the correct place to modify
whether a piece of content is stored in the cache. However, the
documentation recommends the following method used to set uncacheable
content:

  beresp.uncacheable = true[2]

The author suggest this is investigated for static files, rather than
modifying the cache headers of content.

This commit restores the previous 2.1.9 behaviour of setting pages that do
not match a specific whitelist of static files as "uncacheable".

Design notes:

restoration of static files:

  The author feels it's reasonable to cache static files in Varnish.
  While .html is significantly more expensive to generate in Magento
  than static files, which are simply fetched from disk, the content
  is not conceptually different. Additionally, it's reasonable that
  Varnish might be used in graceful failure strategies when upstream
  servers are unavailable, for which it is useful to retain static
  files.

skipped cr:

  This bug was found in production late on a weekday. It skips review,
  and was implemented as a hotfix in the live environment.

[1] - https://www.fastly.com/blog/overriding-origin-ttl-varnish-or-my-beginners-mistake/
[2] - https://book.varnish-software.com/4.0/chapters/VCL_Basics.html#vcl-backend-response
---
 deploy/ansible/templates/etc/varnish/default.vcl | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/deploy/ansible/templates/etc/varnish/default.vcl b/deploy/ansible/templates/etc/varnish/default.vcl
index cfd675d..e6032ea 100644
--- a/deploy/ansible/templates/etc/varnish/default.vcl
+++ b/deploy/ansible/templates/etc/varnish/default.vcl
@@ -186,9 +186,22 @@ sub vcl_backend_response {
     }
 
     # validate if we need to cache it and prevent from setting cookie
+    # SITEWARDS CHANGES START #
+    # Drop cache headers unless resources are static
+    #
+    # See https://github.com/magento/magento2/issues/11939
+    #
+    # images, css and js are cacheable by default so we have to remove cookie also
     if (beresp.ttl > 0s && (bereq.method == "GET" || bereq.method == "HEAD")) {
         unset beresp.http.set-cookie;
+        if (bereq.url !~ "\.(ico|css|js|jpg|jpeg|png|gif|tiff|bmp|gz|tgz|bz2|tbz|mp3|ogg|svg|swf|woff|woff2|eot|ttf|otf)(\?|$)") {
+            set beresp.http.Pragma = "no-cache";
+            set beresp.http.Expires = "-1";
+            set beresp.http.Cache-Control = "no-store, no-cache, must-revalidate, max-age=0";
+            set beresp.grace = 1m;
+        }
     }
+    # SITEWARDS CHANGES END #
 
    # If page is not cacheable then bypass varnish for 2 minutes as Hit-For-Pass
    if (beresp.ttl <= 0s ||
-- 
2.15.0.rc1
```

I can do a PR for this? (but tomorrow lol)

**Comment by andrewhowdencom** [comment]
Oh a quick sanity check shows this caching .html pages also. I'm pretty sure Varnish ignores the headers that are set here, only listens to the headers it receives from upstream *then* allows you to modify them.

Verified by catting `/var/log/nginx/access.log` and noting that there were lots of requests for a .html page from external IP, but only a single request over the loopback for the actual page in PHP over a 5 minute period.

**Comment by magento-engcom-team** [comment]
@andrewhowdencom, thank you for your report.
The issue is already fixed in 2.2.1

**Comment by andrewhowdencom** [comment]
Hola @magento-engcom-team can you point to the commit? I'm looking for it now, but you might save me some work.

**Comment by andrewhowdencom** [comment]
To clarify, I can't see the fix. Absent clarification, please reopen.

**Comment by andrewhowdencom** [comment]
So, there looks to be some sort of special fastly "surrogate control". This would make sense -- specify TTLs for the caching struct separately, so they can't accidentally propagate to the implementing browser. But this isn't how it's setup -- we're using normal cache-control headers, and were stripping them. 

References: 
- https://github.com/magento/magento2/blob/9c14af7a97bcf338662de365d2613c7ca9dbf8c6/lib/internal/Magento/Framework/App/Response/Http.php#L110-L118
- https://github.com/magento/magento2/commit/ce87aa48fff2fdb4ba13bb90fc3630893596c49d
- https://github.com/magento/magento2/search?utf8=%E2%9C%93&q=surrogate&type=

Maybe it's in EE? And didn't make CE. Fastly stuff doesn't have them either; I don't really get how this is all supposed to work.

**Comment by mzeis** [comment]
Let me check if I can find somebody to tell about the status of this issue.

**Comment by mzeis** [comment]
@andrewhowdencom This fix will be in 2.2.2.
